src/main/java/FuncTest/TestFunc/Function.java

```diff
@@ -32,21 +35,64 @@ public HttpResponseMessage run(
 		JSONObject queryResult = new JSONObject(s.getBody().get().toString()).getJSONObject("queryResult");
 		switch (queryResult.getJSONObject("intent").getString("displayName")) {
-      case globals.BUSINESS_HOUR_BY_DAY_INTENT_NAME:
-        return getHourByDay(queryResult, s, c);
-      case globals.BUSINESS_HOUR_WEEK_INTENT_NAME:
-        return getHourByWeek(queryResult, s, c);
-      case globals.FILTER_COURSES_INTENT_NAME:
-        return getMatchingCoursesResponse(queryResult, s, c);
-      case globals.HOMEWORK_GET_UPCOMING_INTENT_NAME:
-        return getUpcomingHomework(queryResult, s, c);
+			case globals.BUSINESS_HOUR_BY_DAY_INTENT_NAME:
+				return getHourByDay(queryResult, s, c);
+			case globals.BUSINESS_HOUR_WEEK_INTENT_NAME:
+				return getHourByWeek(queryResult, s, c);
+			case globals.HOMEWORK_GET_UPCOMING_INTENT_NAME:
+				return getUpcomingHomework(queryResult, s, c);
+			case globals.LOGIN_INTENT:
+				return checkLoginName(queryResult, s, c);
+			case globals.SUBSCRIBE_INTENT:
+				return subscribeToSystem(queryResult,s,c);
+			case globals.FILTER_COURSES_INTENT_NAME:
+				return getMatchingCoursesResponse(queryResult, s, c);
+			case globals.RUN_RULES_INTENT:
+				return applyRules(queryResult, s, c);
       case globals.COURSE_GET_POSTREQUISITES_BY_NAME_INTENT_NAME:
          return PostrequisiteHandler.getPostrequisitesByName(queryResult, s, c);
       case globals.COURSE_GET_POSTREQUISITES_BY_NUMBER_INTENT_NAME:
         return PostrequisiteHandler.getPostrequisitesByNumber(queryResult, s, c);
 		}
 		return utils.createWebhookResponseContent("what is this intent?", s);
 	}
+	
+	private HttpResponseMessage applyRules(JSONObject queryResult, HttpRequestMessage<Optional<String>> s,
+			ExecutionContext c) {
+		// TODO Auto-generated method stub
+		String uname = utils.getStringUserParamFromContext(queryResult, "username");
+		String passwd = utils.getStringUserParamFromContext(queryResult, "password");
+		
+		c.getLogger().info("================ username = " + uname + "================");
+		c.getLogger().info("================ password = " + passwd + "================");
+		
+		return new RunRulesHandler(uname, passwd).runHomeworkRules(s,c);
+	}
+
+	private HttpResponseMessage subscribeToSystem(JSONObject queryResult, HttpRequestMessage<Optional<String>> s,
+			final ExecutionContext c) {
+		JSONObject parameters = queryResult.getJSONObject("parameters");
+		List<String> requiredParameterNames = Arrays.asList("username", "password");
+		
+		if (!utils.allParametersArePresent(parameters, requiredParameterNames)) 
+			return utils.createWebhookResponseContent("Please enter user credentials to subscribe with.", s);
+		
+		return utils.createWebhookResponseContent(
+				(new subscribeHandler(parameters.getString("username"),parameters.getString("password"))).subscribe(c), s);
+	}
+	
+	private HttpResponseMessage checkLoginName(JSONObject queryResult, HttpRequestMessage<Optional<String>> s,
```

Comment:
Duplicate of the previous function, so it seems to be redundant.

Reply:
I disagree with that. The former checks the user has an entered details and then performs a "subscription" (writes to the DB). The latter, check the user has entered details and then checks they are correct (reads from the DB).
Both operations are done through intermediary classes.

I believe though they are similar each possesses a distinct "source of truth" (as we learned in Software Design) and so it's fine in my opinion

Reply:
I do not agree as explained in the conversation under the "files changed" tab